### PR TITLE
feat: add hashkey pattern rule

### PR DESCRIPTION
**Pull request by levy5307**
### What problem does this PR solve? 
RFC: https://github.com/apache/incubator-pegasus/blob/master/rfcs/2021-05-27-user-specified-compaction.md

##### Tests 

- Unit test

### CODE REVIEW

File: src/server/compaction_filter_rule.h [hunk 1]
```diff
@@ -36,5 +38,34 @@ class compaction_filter_rule
...
+ENUM_BEGIN(string_match_type, SMT_INVALID)
+ENUM_REG(SMT_MATCH_ANYWHERE)
+ENUM_REG(SMT_MATCH_PREFIX)
+ENUM_REG(SMT_MATCH_POSTFIX)
+ENUM_END(string_match_type)
+
+class hashkey_pattern_rule : public compaction_filter_rule
+{
+public:
+    explicit hashkey_pattern_rule(uint32_t pegasus_data_version);
```
Comment (hycdong) [inline src/server/compaction_filter_rule.h:58]: Why do hashkey pattern need `pegasus_data_version`? Is `pegasus_data_version` used for value schema?

File: src/server/compaction_filter_rule.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+namespace pegasus {
+namespace server {
+bool string_pattern_match(const std::string &value,
+                          string_match_type type,
+                          const std::string &filter_pattern)
+{
+    if (filter_pattern.length() == 0)
+        return false;
+    if (value.length() < filter_pattern.length())
+        return false;
```
Comment (acelyc111) [inline src/server/compaction_filter_rule.cpp:35]:
```suggestion
    if (filter_pattern.empty()) {
        return false;
    }
    if (value.length() < filter_pattern.length()) {
        return false;
    }
```

Comment (acelyc111) [inline src/server/compaction_filter_rule.cpp:35]: UT doesn't cover this case.

File: src/server/compaction_filter_rule.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    case string_match_type::SMT_MATCH_ANYWHERE:
+        return dsn::string_view(value).find(filter_pattern) != dsn::string_view::npos;
+    case string_match_type::SMT_MATCH_PREFIX:
+        return memcmp(value.data(), filter_pattern.data(), filter_pattern.length()) == 0;
+    case string_match_type::SMT_MATCH_POSTFIX:
+        return memcmp(value.data() + value.length() - filter_pattern.length(),
+                      filter_pattern.data(),
+                      filter_pattern.length()) == 0;
+    default:
+        derror_f("invalid match type {}", type);
```
Comment (acelyc111) [inline src/server/compaction_filter_rule.cpp:47]: UT doesn't cover this case.

File: src/server/compaction_filter_rule.h (hunk 1 above)
Comment (levy5307) [reply src/server/compaction_filter_rule.h:58]: Yes, I can remove it in this pull request.

**Review by hycdong — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]
LGTM

**Pull request merged**